migrations/20231128_jellyfish_migration/utils/jellyfish_data.go

```diff
@@ -0,0 +1,165 @@
+package utils
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/tidepool-org/platform/pointer"
+)
+
+func jellyfishQuery(settings Settings, userID *string, lastFetchedID *string) (bson.M, *options.FindOptions) {
+	selector := bson.M{
+		"_deduplicator": bson.M{"$exists": false},
+	}
+	if settings.Rollback {
+		selector = bson.M{
+			settings.RollbackSectionName: bson.M{"$exists": true},
+		}
+	}
+
+	if userID != nil && *userID != "" {
+		log.Printf("fetching for user %s", *userID)
+		selector["_userId"] = *userID
+	}
+	idNotObjectID := bson.M{"$not": bson.M{"$type": "objectId"}}
+
+	if lastFetchedID != nil && *lastFetchedID != "" {
+		selector["$and"] = []interface{}{
+			bson.M{"_id": bson.M{"$gt": *lastFetchedID}},
+			bson.M{"_id": idNotObjectID},
+		}
+	} else {
+		selector["_id"] = idNotObjectID
+	}
+
+	bSize := int32(settings.QueryBatchSize)
+	limit := int64(settings.QueryBatchLimit)
+	opts := &options.FindOptions{
+		Sort:      bson.M{"_id": 1},
+		BatchSize: &bSize,
+		Limit:     &limit,
+	}
+
+	return selector, opts
+}
+
+var JellyfishDataQueryFn = func(m *DataMigration) bool {
+
+	settings := m.GetSettings()
+
+	if dataC := m.GetDataCollection(); dataC != nil {
+
+		selector, opts := jellyfishQuery(
+			settings,
+			nil,
+			pointer.FromString(m.GetLastID()),
+		)
+
+		dDataCursor, err := dataC.Find(m.GetCtx(), selector, opts)
+		if err != nil {
+			log.Printf("failed to select data: %s", err)
+			return false
+		}
+		defer dDataCursor.Close(m.GetCtx())
+
+		all := []bson.M{}
+
+		for dDataCursor.Next(m.GetCtx()) {
+			item := bson.M{}
+			if err := dDataCursor.Decode(&item); err != nil {
+				log.Printf("error decoding data: %s", err)
+				return false
+			}
+			itemID := fmt.Sprintf("%v", item["_id"])
+			userID := fmt.Sprintf("%v", item["_userId"])
+			itemType := fmt.Sprintf("%v", item["type"])
+			if settings.Rollback {
+				if rollback, ok := item[settings.RollbackSectionName].(primitive.A); ok {
+					cmds := []bson.M{}
+					for _, cmd := range rollback {
+						if cmd, ok := cmd.(bson.M); ok {
+							cmds = append(cmds, cmd)
+						}
+					}
+					if len(cmds) > 0 {
+						m.SetUpdates(UpdateData{
+							Filter:   bson.M{"_id": itemID, "modifiedTime": item["modifiedTime"]},
```

Comment:
Modified time is so we don't stop on other changes happening simultaneously-ish?

Reply:
So, I guess if this happens ad an update is excluded because it changed by something else, we'd pick this up on the next go around?